### PR TITLE
upgrade and call

### PR DESCRIPTION
**Pull request by juniset**
This PR aims to enable to upgrade the account and execute a multicall to the new implementation in the same transaction.

### CODE REVIEW

File: test/proxy.py [hunk 1]
```diff
@@ -3,26 +3,34 @@
...
+from utils.utilities import compile, cached_contract, assert_revert, assert_event_emmited, get_execute_data
+from utils.TransactionSender import TransactionSender, from_call_to_call_array
+from typing import Optional, List, Tuple
+
 from starkware.starknet.compiler.compile import get_selector_from_name
 
 signer = Signer(1)
 guardian = Signer(2)
 wrong_signer = Signer(3)
 wrong_guardian = Signer(4)
```
Comment (delaaxe) [inline test/proxy.py:16]: You could define a type for calls, either here or in a shared file:

```python
Call = Tuple[str, str, List]
```

File: test/proxy.py [hunk 2]
```diff
@@ -81,34 +100,194 @@ async def test_call_dapp(contract_factory):
     await sender.send_transaction([(dapp.contract_address, 'set_number', [47])], [signer, guardian])
     assert (await dapp.get_number(account.contract_address).call()).result.number == 47
 
+
+def build_execute_after_upgrade_data(calls: Optional[List[Tuple]] = None) -> List[Tuple]:
```
Comment (delaaxe) [inline test/proxy.py:104]: And then use it in lieu of `Tuple`:

```suggestion
def build_execute_after_upgrade_data(calls: Optional[List[Call]] = None) -> List[Tuple]:
```

Comment (delaaxe) [inline test/proxy.py:104]: Return type seems to be `List[Union[int, Tuple]]`

File: test/proxy.py [hunk 3]
```diff
@@ -81,34 +100,194 @@ async def test_call_dapp(contract_factory):
...
+        len(multicall_calldata),
+        *multicall_calldata
+    ]
+
+
+def build_upgrade_call(
+        account,
+        new_implementation,
+        calls: Optional[List[Tuple]] = None,
+) -> Tuple:
```
Comment (delaaxe) [inline test/proxy.py:122]:
```suggestion
        calls: Optional[List[Call]] = None,
) -> Call:
```

File: test/utils/TransactionSender.py [hunk 4]
```diff
@@ -11,17 +11,18 @@
...
-class TransactionSender():
+
+class TransactionSender:
     def __init__(self, account: StarknetContract):
         self.account = account
 
     async def send_transaction(
         self,
-        calls,
+        calls: List,
```
Comment (delaaxe) [inline test/utils/TransactionSender.py:21]:
```suggestion
        calls: List[Call],
```

File: test/utils/TransactionSender.py [hunk 5]
```diff
@@ -95,7 +96,8 @@ async def declare_class(
         execution_info = await state.execute_tx(tx=tx)
         return execution_info
 
-def from_call_to_call_array(calls):
+
+def from_call_to_call_array(calls: List):
```
Comment (delaaxe) [inline test/utils/TransactionSender.py:100]: etc

File: test/utils/utilities.py [hunk 6]
```diff
@@ -41,7 +43,7 @@ def assert_event_emmited(tx_exec_info: TransactionExecutionInfo, from_address: i
         from_address=from_address,
         keys=[get_selector_from_name(name)],
         data=data,
-    ) in raw_events
+    ) in raw_events, f"Event {name} not found"
```
Comment (delaaxe) [inline test/utils/utilities.py:46]: Maybe I'd introduce a boolean variable here to make it a bit more readable

File: test/utils/utilities.py [hunk 7]
```diff
@@ -70,4 +72,11 @@ def cached_contract(state: StarknetState, _class: ContractClass, deployed: Stark
         contract_address=deployed.contract_address,
         deploy_call_info=deployed.deploy_call_info
     )
-    return contract
\ No newline at end of file
+    return contract
+
+
+def get_execute_data(tx_exec_info: TransactionExecutionInfo) -> List[int]:
+    raw_data:List[int] = tx_exec_info.call_info.retdata
```
Comment (delaaxe) [inline test/utils/utilities.py:79]:
```suggestion
    raw_data: List[int] = tx_exec_info.call_info.retdata
```

File: contracts/account/ArgentAccount.cairo [hunk 8]
```diff
@@ -213,10 +210,38 @@ func initialize{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}
...
-    implementation: felt
-) {
+    implementation: felt, calldata_len: felt, calldata: felt*
+) -> (retdata_len: felt, retdata: felt*) {
+    alloc_locals;
     ArgentModel.upgrade(implementation);
-    return ();
+
+    if (calldata_len == 0) {
+        local retdata : felt*;
```
Comment (juniset) [inline contracts/account/ArgentAccount.cairo:219]: I think you don't need to specify that `retdata` is local (`let retdata` is sufficient). This is taken care of by the `alloc_locals` you have at the beginning of the method.

**Review by sgc-code — APPROVED** [review]
Lgtm

File: test/proxy.py (hunk 1 above)
Comment (sgc-code) [reply test/proxy.py:16]: i agree with all of the comments, we had to merge this PR a bit quicker but i will implement the changes next week 👍

**Pull request merged**